Issue 1160

kaiiam opened the pull request:
@pbuttigieg @rduerr PR addressing https://github.com/EnvironmentOntology/envo/issues/1160 last batch of cryospheric work.

kaiiam commented:
@rduerr @pbuttigieg @brandonnodnarb

Build finally passed, was having issues with obsoleting the old `cryosphere` term which was causing errors with the travis build. It's settled now. There were some issues with duplicate ID's in the template, and I've sorted those as well. I think this is ready for review.

kaiiam commented:
I realize I need to pad Ruth's ID's `ENVO:3400003` to `ENVO:3400003`

kaiiam commented:
I've fixed the IDs and the build passes. I think we're good now, new and substantially modified terms resulting from this PR are as follows: 

- [ ] ENVO:00000429
- [ ] ENVO:00000484
- [ ] ENVO:01000818
- [ ] ENVO:01000880
- [ ] ENVO:01000882
- [ ] ENVO:01000895
- [ ] ENVO:01000896
- [ ] ENVO:03000116
- [ ] ENVO:03000141
- [ ] ENVO:03000142
- [ ] ENVO:03000143
- [ ] ENVO:03000144
- [ ] ENVO:03000145
- [ ] ENVO:09200020
- [ ] ENVO:03400003
- [ ] ENVO:03400004
- [ ] ENVO:03400005
- [ ] ENVO:03400006
- [ ] ENVO:03400007
- [ ] ENVO:03400008
- [ ] ENVO:03400009
- [ ] ENVO:03400010
- [ ] ENVO:03400011
- [ ] ENVO:03400012
- [ ] ENVO:03400013
- [ ] ENVO:03400014
- [ ] ENVO:03400015
- [ ] ENVO:03400016
- [ ] ENVO:03400017

Please review when you get the chance.

wdduncan commented:
For the cloud clases (e.g.,`altocumulus cloud`), I see axioms like:
```
 'composed primarily of' some ('object aggregate' and ('has member' some 'water droplet'))
```
and
```
'composed primarily of' some 'liquid water'
```
Should these axioms belong to the parent `water-based cloud`?

Aslo (speaking tersely) if a cloud is primarily composed of water droplets (an aggregate), wouldn't it also be primarily composed of water (cf. class `water droplet`)? In other words, can the axiom `'composed primarily of' some 'liquid water'` be eliminated?

wdduncan commented:
The definition for `hail` reads: `Precipitation which is composed primarily of hail stones.`

Should it be: `An object aggregate which is composed primarily of hail stones.`

Also, for `falling sleet`, `falling snow`, and `hail` do you want to add the axiom?:
```
'participates in' some 'precipitation process'
```

wdduncan commented:
Did you want make `cryosphere` a defined class? It seems like it might be a candidate for this, but I'm not sure.

raissameyer commented:
> For the cloud clases (e.g.,`altocumulus cloud`), I see axioms like:
> ```
> 'composed primarily of' some ('object aggregate' and ('has member' some 'water droplet'))
> ```
> and
> ```
> 'composed primarily of' some 'liquid water'
> ```
> Should these axioms belong to the parent `water-based cloud`?

@wdduncan We had kept those axioms on the subclass level, as they are not the same throughout all water-based clouds, e.g.,
* altocumulus cloud: `'composed primarily of' some 
    ('object aggregate'
     and ('has member' some 'water droplet'))` and `'composed primarily of' some 'liquid water'`
* cirrocumulus cloud: `'composed primarily of' some 
    ('object aggregate'
     and ('has member' some 'water ice crystal'))` and `'composed primarily of' some 'water ice'`

We do, however, have the more general axiom `'composed primarily of' some 
    ('liquid water' or 'water ice')` on the "water-based cloud"

> Aslo (speaking tersely) if a cloud is primarily composed of water droplets (an aggregate), wouldn't it also be primarily composed of water (cf. class `water droplet`)? In other words, can the axiom `'composed primarily of' some 'liquid water'` be eliminated?

As "water droplet" specifies that it is a `droplet
 and ('composed primarily of' some 'liquid water')` I think that we could indeed eliminate that axiom. 

@pbuttigieg Do you remember the reason as to why we additionally added the `'composed primarily of' some 'liquid water'` / `'composed primarily of' some 'water ice'` axioms?

cmungall commented:
I agree with @wdduncan's comments re hail.

But more broadly I have some concerns. For the general concept of 'hail' this PR introduces 3 classes:

 1. hail _**Precipitation** which is composed primarily of **hail stones**_
 2. hailstone _An ice mass which is in a ball-shape or irregular lump, with a diameter ranging from 5 to 50 millimeters (0.2 to 2.0 inches) or more, always produced by convective clouds, nearly always cumulonimbus_
 3. hailstorm _An atmospheric storm which intense **hailfall** occurs._

I have concerns specifically with 1

It has a text definition of “Precipitation which is composed primarily of hail stones”. This is a problem as the genus “precipitation”
 - (a) is not a class in ENVO and
 - (b) it is suggestive of the process (which is in -- http://purl.obolibrary.org/obo/ENVO_01000875) and a process cannot be the domain of compose-primarily-of

This logical inconsistency is hidden (I call these cryptic incoherencies) because of the lack of logical axioms on the class. 

In fact the only logical axiom on "hail" is a subclassof to 'BFO' object aggregate
SubClassOf(&lt;http://purl.obolibrary.org/obo/ENVO_03400011&gt; &lt;http://purl.obolibrary.org/obo/BFO_0000027&gt;)

 - Textual definitions should match logical axioms to a reasonable extent.
 - classes should be reasonably interlinked - currently the class for 'hail' is not linked to the class 'hailstone'

If there is truly a use case for introducing a class for 'hail' as distinct from hailstone, then I would introduce a new class for the parent (something like "portion of precipitation" -- this sounds horrible but I think it reflects the artificial nature of this class), and a defining axiom `= portion of precipitation and composed-primarily-of some hailstone'. I would also define a design pattern in yaml and make the structure of other kinds of precipitation.

I also see an xref to:
http://sweetontology.net/phenAtmoPrecipitation/FreezingRain

Is this deliberate? This is already xreffed to ENVO_03400010 'freezing rain', which is more appropriate.

I recommend using skos not oio, having an exactMatch between sweet:FreezinRain and ENVO freezing rain; if you feel th SWEET concept does need mapped to "hail" as well, then use closeMatch.

Minor: I also see 8 createdBys on this one term - is this really the case?

Overall I would recommend keeping things simple. I see a lot of axioms with nested concepts. These are harder for humans to reason about, and frequently they mean different things in OWL than you might think. They are a frequent cause of cryptic incoherency. I also see missing axioms - I'd expect to see more linkages between some of these concepts.

I would also state the Less is More dictum. Creating similar-sound classes such as 'hail' and hailstone' is justifiable using the BFO framework, but you are introducing potential sources of ambiguity and annotator inconsistency. When you do introduce these shadow concepts, be sure to include curator to notes to help indicate which one should be used in annotation in any give context. And make things more predictable by following design patterns, so that we see the same structures, e.g. rain droplet vs rain vs rain precipiration, snow vs snowflake vs snow precipitation.

cmungall commented:
A general note on the pattern:

`composed primarily of' some ('object aggregate' and ('has member' some X))`

This may be weaker than you intend. In fact it is no stronger than saying has part. An object aggregate can have many members.

What exactly is the use case for these axioms? What inferences are you hoping to get from them? Or do they exist primarily for humans to navigate, or to support certain kinds of queries?

pbuttigieg commented:
@kaiiam thanks for prepping the PR, the ROBOT template approach seems to be working quite well.

@wdduncan 
>The definition for hail reads: Precipitation which is composed primarily of hail stones.
>
> Should it be: An object aggregate which is composed primarily of hail stones.

Yes, it can be - hail on the ground is still hail, but the semantic looseness around "precipitation" would allow for that too.

> Also, for falling sleet, falling snow, and hail do you want to add the axiom?:
>
>'participates in' some 'precipitation process'

Yes, thanks - that will help link things

pbuttigieg commented:
@wdduncan thanks for the feedback on the cloud compositions. As @raissameyer noted, there are clouds primarily composed of ice, thus the axioms.

>  Aslo (speaking tersely) if a cloud is primarily composed of water droplets (an aggregate), wouldn't it also be primarily composed of water (cf. class water droplet)? In other words, can the axiom 'composed primarily of' some 'liquid water' be eliminated?
>
>As "water droplet" specifies that it is a droplet and ('composed primarily of' some 'liquid water') I think that we could indeed eliminate that axiom.
>
> @pbuttigieg Do you remember the reason as to why we additionally added the 'composed primarily of' some 'liquid water' / 'composed primarily of' some 'water ice' axioms?

@raissameyer I think that was an oversight - @wdduncan's analysis and yours are correct.

pbuttigieg commented:
@cmungall 

Yes, we do need hail (the object aggregate) in addition to hailstone (a member of the aggregate). @wdduncan's suggested definition should fix that. 

I think the mapping to SWEET freezing rain is erroneous. 


> Minor: I also see 8 createdBys on this one term - is this really the case?

Yes, there were eight creators as this comes out of our Cryohackathon and we're working collectively. 


>I recommend using skos not oio, having an exactMatch between sweet:FreezinRain and ENVO freezing rain; if you feel th SWEET concept does need mapped to "hail" as well, then use closeMatch.

The tacit clarification here (@raissameyer relevant for your recent wiki page) is that SKOS APs for synonyms should point to URIs, whereas the OIO (oboInOwl) APs can point to literals. I don't think we have SKOS APs in ENVO yet, but this is a very good chance to add them and start a broader clean up.


> Overall I would recommend keeping things simple. I see a lot of axioms with nested concepts. These are harder for humans to reason about, and frequently they mean different things in OWL than you might think.

We're capturing the knowledge of domain experts in this series of Cryohackathons - clean up for performance and OWL surprises can certainly be a follow up mission. 

> I also see missing axioms - I'd expect to see more linkages between some of these concepts.

Absolutely, keep in mind our time is tight during these working sessions. If you see more opportunities, please open an issue or a PR.

> I would also state the Less is More dictum. Creating similar-sound classes such as 'hail' and hailstone' is justifiable using the BFO framework, but you are introducing potential sources of ambiguity and annotator inconsistency.

This is not driven by BFO, but the need to unpack these from @rduerr's synthesis of >20 Global Cryosphere Watch glossaries: the confusion is real.


> When you do introduce these shadow concepts, be sure to include curator to notes to help indicate which one should be used in annotation in any give context. 

They are not shadowing each other, they are very real; however, this is a very good suggestion and we should create a wiki page about handling closely related concepts with this as an example.


> And make things more predictable by following design patterns, so that we see the same structures, e.g. rain droplet vs rain vs rain precipiration, snow vs snowflake vs snow precipitation.

Yes, we're discovering these post hoc - first we have to capture the raw knowledge and see if things can be made into patterns. There are often hidden subtleties which have to be reflected on.

kaiiam commented:
Thanks all for the feedback, the discussion the cloud semantics is good and interesting but was not part of the terms included in this PR. 

One issue I noted is that ID ENVO:09200020 is in ENVO as  `concentration of nitrate in groundwater` (which appears not to have been built using the chemical concentration DOSDP pattern), but that ID is also used in the entity_attribute csv: https://github.com/EnvironmentOntology/envo/blob/master/src/envo/modules/entity_attribute.csv#L18 as depth of snow, but it doesn't appear to have been compiled from there. We'll need to fix that ID in the entity_attribute csv.

kaiiam commented:
Looking at this again, I'm a bit confused about the [entity_attribute.csv](https://github.com/EnvironmentOntology/envo/blob/master/src/envo/modules/entity_attribute.csv) table the terms:

`ENVO:09200019` and `ENVO:09200020` are not compiled but the rest are compiled. :shug: I don't understand what's up here but that's getting in the way here as we wanted to use `ENVO:09200020` as depth of snow which it's suppose to be in that DOSDP table. However confounding this the term is already complied as `concentration of nitrate in groundwater` not from a design pattern but was generated manually in protege. Since `ENVO:09200020` is already out there as nitrate I think we should keep that there and change `depth of snow` away from `ENVO:09200020` to a new ID. same with `ENVO:09200019` which is suppose to be `mass density of snow` I therefore propose we fix this by making the following:

I propose we fix this by changing:

`ENVO:09200019` to `ENVO:3100035` 
`ENVO:09200020` to `ENVO:3100036` 
 
@pbuttigieg reasonable?

pbuttigieg commented:
@kaiiam 

Yes, reasonable - we shouldn't change any IRIs that are out in the wild post-release. It is odd that this happened, but non-patterned terms will show up from time to time.

kaiiam commented:
I've now updated this with the final changes. This is now ready for final review @pbuttigieg @rduerr @wdduncan. Thanks to the ESIP semantic harmonization committee for being patient here.

kaiiam commented:
@wiegandn you can also see and if you'd like review the changes here for now.

kaiiam commented:
@cmungall @wdduncan @pbuttigieg we're getting the following build error. My guess is that our new build process without travis doesn't have as sufficient memory to run the make the ncbitaxon import? I think we need to sort this out as this will probably be affecting all PRs (unless I'm missing something here). 

```
make: *** [Makefile:410: mirror/ncbitaxon.owl] Error 137
rm imports/foodon_combined_seed.tsv imports/chebi_combined_seed.tsv imports/pato_combined_seed.tsv imports/go_combined_seed.tsv imports/pco_combined_seed.tsv imports/ro_combined_seed.tsv imports/uberon_combined_seed.tsv
Error: Process completed with exit code 2.
```

kaiiam commented:
It's the same error as in https://github.com/EnvironmentOntology/envo/pull/1163 I think we've got a build problem.

wdduncan commented:
@kaiiam Yes. There is a problem with not having enough memory. For the recent merges, I ran `make test` locally to make sure everything was okay.  

Are you able to do this locally? If not, I can checkout the branch and run on my machine.

cmungall commented:
Getting actions working is higher priority than definition tweaking

All we have to do: **don't build `all_imports` as part of tests**

Recall that when we were using travis we made this target:

```
travis_test: test_python envo-edit.owl all_reports all_modules envo.owl
```

This differs from the main test target in that it doesn't include `all_imports`

The switch from travis to actions is purely a devops one, our test logic should be identical

wdduncan commented:
@kaiiam The memory problem has been addressed by @cmungall (Thanks Chris!).  
The PR still fails due some QC checks; e.g. see [previous comment](https://github.com/EnvironmentOntology/envo/pull/1172#issuecomment-1035206878).

kaiiam commented:
@cmungall you rock thanks for fixing this in https://github.com/EnvironmentOntology/envo/commit/4238f50caf1ad005c83226cdb08feb72a2f197da! Its great that CI quality check build script is back! 

Also @wdduncan thank you identifying the duplicate defs and merging master into this branch. I'll fix up those defs.

kaiiam commented:
Build is now passing and ready to merge @pbuttigieg @rduerr @wiegandn please have a look or let me know if we're good to merge!

rduerr commented:
@kaiiam I'd love to review this but if you have suggestions for exactly how I should go about doing that, that would be helpful.  I have loaded what I think is the #1172 branch of envo-edit into protege and it at least is fine per elk; and I did look at the diff files though the last two seem very opaque to me since I couldn't even find those terms in the envo-edit file.  However, it isn't clear how to validate each of the changes made, unless it is by looking at each individual commit and the terms in each in protege?  Is that what I am supposed to do?  If so, then don't expect something real soon.... since that will take hours....

kaiiam commented:
@rduerr I agree reviewing OWL on github  can be a pain. However, this PR comes from sheet 2 of [ESIP_Cryo_hackathon_robot_template](https://docs.google.com/spreadsheets/d/1Xt3BHNz-pP2ggJq_asPPcuKuGEAKtbT3MU2wFLTnCmQ/edit?usp=sharing), hence we have the list of IDs added. Its 24 long so I think that's reasonable to double check. One can simply copy these IDs and then in protege with the branch's envo-edit.owl file loaded hit `control (or cmd on mac) + F`. Then paste the id in the search box then hit enter. That'll take you to the term. 

<img width="269" alt="image" src="https://user-images.githubusercontent.com/12255688/155086452-959a65dc-0138-444b-a8ae-c0bd7262b6ab.png">

I've checked them and think it's correct with what we put in the robot template (with all the mistakes I found fixed). If there is anything from https://github.com/EnvironmentOntology/envo/issues/1291 you'd like to change please let me know and I can update this PR, otherwise I'd say to just check for obvious errors/mistakes so we can merge these changes in.

```
ENVO:03000143
ENVO:03000144
ENVO:00000484
ENVO:03000141
ENVO:03000145
ENVO:03000142
ENVO:03000116
ENVO:03400003
ENVO:09200020
ENVO:01000896
ENVO:03400010
ENVO:03400011
ENVO:03400012
ENVO:03400013
ENVO:01000895
ENVO:01000880
ENVO:03400004
ENVO:03400005
ENVO:01000882
ENVO:03400006
ENVO:03400007
ENVO:03400008
ENVO:03400009
ENVO:00000429
```

kaiiam commented:
@pbuttigieg @rduerr and I will pick this up again next week from `ENVO:03400012`.

kaiiam commented:
I'd like to fold addressing the issues in https://github.com/EnvironmentOntology/envo/issues/1069 into this PR, upon urgent  request from @matentzn, fixing ENVO_01001784 (which chris suggests to remove the EQ axiom) comparing against ENVO_01001479.

pbuttigieg commented:
fixing some reasoner errors detected by hermit 
![image](https://user-images.githubusercontent.com/2276146/158674783-8f4966a4-0fed-4341-96fe-1c66128526c9.png)

fossil fuel and and biological pigment are roles in CHEBI, not material entities.

 @matentzn of interest to you and @laurenechan
